tajo-plan/src/main/java/org/apache/tajo/plan/joinorder/GreedyHeuristicJoinOrderAlgorithm.java

```diff
@@ -57,17 +54,69 @@ public FoundJoinOrder findBestOrder(LogicalPlan plan, LogicalPlan.QueryBlock blo
     JoinEdge bestPair;
 
     while (remainRelations.size() > 1) {
+      Set<LogicalNode> checkingRelations = new LinkedHashSet<LogicalNode>();
+
+      for (LogicalNode relation : remainRelations) {
+        Collection <String> relationStrings = PlannerUtil.getRelationLineageWithinQueryBlock(plan, relation);
+        List<JoinEdge> joinEdges = new ArrayList<JoinEdge>();
+        String relationCollection = TUtil.collectionToString(relationStrings, ",");
+        List<JoinEdge> joinEdgesForGiven = joinGraph.getIncomingEdges(relationCollection);
+        if (joinEdgesForGiven != null) {
+          joinEdges.addAll(joinEdgesForGiven);
+        }
+        for (String relationString: relationStrings) {
```

Comment:
Oh, it's my mistake.
When relationStrings has only one entry, this code may adds that entry twice.

When a LogicalNode contains two relations, for example, A and B,
above code first finds joinEdges whose right relation is "A, B", which is obtained by TUtil.collectionToString().
Next, it finds joinEdges whose right relation is "A" or "B", which is obtained by 'for (String relationString: relationStrings)'.
So, if a LogicalNode contains just one relation, this code may adds that relation repeatedly.

Duplicated relation does not affect the result but I'll patch not to have duplicated relations.